### PR TITLE
Shippo plugin

### PR DESCRIPTION
**Pull request by prosf**
New Shippo package for #421 

## Features:
1. Shippo added as plugin with API key to connect
2. Shippo Dashboard UI allows for carriers to be enabled/disabled (using existing Shipping providers table)
3. Shippo is integrated in the CartFlow, gets rates for enabled carriers and allows for buyers to select a shipping method, total is updated, and selected Shippo costs are saved in the current order
4. Shippo carriers and costs are displayed in all orders
6. Selected Shipping provider with selected method (ie. overnight) is shown in all confirmation screens
7. The Shippo order is confirmed upon seller capturing the payment from the dashboard
8. Tracking number returned from Shippo upon capturing payment is set in current order 
9. Shipping Labels can be downloaded from the order summary directly when payment is captured
10. **EDIT** We added a `Fetch Carriers` button to get only `active` carriers from Shippo. 
The way Shippo carriers work in RC is the following:
    - When API Key is added - carriers are fetched and saved in RC as shipping docs 
    - Admin is then able to choose which carriers will be displayed in the CartFlow as choices (saves us from fetching carriers all the time or setting up a job for something used very rarely)
    - User might go in Shippo and enable/disable another carrier, Fetch will be used to update RC

Notes:
- Currently the item has its dimensions and weight values taken from Products.json

## In Review:
- [x] Tracking status updates from Shippo (from the time of order - up until delivery) - We're currently developing this through a Job - Shippo updates every two hours, we'll run a job for that as well and submit a different PR to this one.

## ToDo:
- [ ] International orders work for DHL, but need to be looked for USPS, FedEx, and possibly others, so that all information is included for the customs label https://goshippo.com/docs/international.
- [ ] This requires the product schema to be extended with fields related to Customs (ie. item_description, country_of_origin, etc ) https://goshippo.com/docs/reference#customs-items
- [ ] Change default UnitsOfMeasure to also include distance along with weight for each shop
- [ ] UI for parcel details (ie. weight, dimensions)
- [ ] UI for Customs fields https://goshippo.com/docs/reference#parcels
- [ ] Create more meaningful error messages for all ShippoAPI errors
- [ ] Check ShippoAPI responses for errors given by Shippo, when the requests are successful (Check response status/state)
- [ ] Multiple items from same seller have not been implemented

### CODE REVIEW

**Comment by zenweasel** [comment]
So I am on this branch and have added an API key and got a success message but I am not seeing the shipping method showing up during checkout.

cart_checkout

**Comment by aaronjudd** [comment]
I think it'd be a good idea to follow the format that we just implemented in [taxes and payments](https://github.com/reactioncommerce/reaction/tree/release-0.18.0/imports/plugins/core/payments/client/settings) with settings, we'd store the enabled method settings something like:

```
 "enabled": true, 
 "settings" : {
        "shippo": {
            enabled: true
        },
      "api_key": "xx"
```

If one of the templates is defined in `register.js` as ```provides: shippingMethod``` this registry / settings configuration will control the visibility of the shipping methods available in the checkout.  {reactionApps provides="shippingMethod"} is where the client shipping methods should be rendered (where @zenweasel reported the "no shipping methods").

It was my hope that we'd be able to update shipping to follow this format as a standard in this release or 0.19.

File: imports/plugins/included/shippo/register.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      label: "Shippo",
+      permission: "dashboard/shippo"
+    }]
+  }, {
+    label: "Shippo",
+    route: "/dashboard/shippo",
+    provides: "settings",
+    container: "connection",
+    template: "shippoSettings"
+  }
```
Comment (aaronjudd) [inline imports/plugins/included/shippo/register.js]: Add `provides:shippingMethod` here to render checkout shipping methods.

File: imports/plugins/included/payments-stripe/client/settings/stripe.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ Template.stripeSettings.helpers({
     return StripePackageConfig;
   },
   packageData() {
+
```
Comment (zenweasel) [inline imports/plugins/included/payments-stripe/client/settings/stripe.js]: Did we mean to add a blank line to an unrelated package?

File: imports/plugins/included/shippo/register.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { Reaction } from "/server/api";
+
+Reaction.registerPackage({
+  label: "Shippo",
+  name: "reaction-shippo",
+  icon: "fa fa-plane",
+  autoEnable: true,
+  settings: {
+    api_key: ""
+  },
```
Comment (zenweasel) [inline imports/plugins/included/shippo/register.js]: key should be camel-cased

File: imports/plugins/included/shippo/register.js (hunk 1 above)
Comment (zenweasel) [inline imports/plugins/included/shippo/register.js]: Don't we need something like this block? (i.e. we need a checkout template as well)

```
    {
      template: "flatRateCheckoutShipping",
      name: "shipping/flatRates",
      provides: "shippingMethod"
    }
```

**Comment by zenweasel** [comment]
It looks like there's a lot of pieces not implemented like methods for fetching rates, etc. I'm going to mark this WIP while get that stuff finished.

File: imports/plugins/included/shippo/register.js [hunk 4]
```diff
@@ -25,6 +28,10 @@ Reaction.registerPackage({
     provides: "settings",
     container: "connection",
     template: "shippoSettings"
+  }, {
+    provides: "shippingMethod",
+    name: "shipping/methods/shippo",
+    template: "shippoMethods"
```
Comment (zenweasel) [inline imports/plugins/included/shippo/register.js]: Is this template available somewhere?

**Comment by lcampanis** [comment]
@aaronjudd Shippo is now updated and ready to be reviewed. There's a few things that need to be looked at. So best you guys review it and we can start working on things together ;)

**Comment by zenweasel** [comment]
So when I go to checkout the second time on a logged in user I am getting a string of errors and shipping methods never load:

```
Exception from Tracker recompute function:
meteor.js?hash=e3f53db…:930 Error: No such template: spinner
    at blaze.js?hash=9d08dd7…:3214
    at Blaze.View.<anonymous> (spacebars.js?hash=ebf9381…:68)
    at blaze.js?hash=9d08dd7…:1934
    at Function.Template._withTemplateInstanceFunc (blaze.js?hash=9d08dd7…:3746)
    at blaze.js?hash=9d08dd7…:1932
    at Object.Blaze._withCurrentView (blaze.js?hash=9d08dd7…:2273)
    at viewAutorun (blaze.js?hash=9d08dd7…:1931)
    at Tracker.Computation._compute (tracker.js?hash=9f8a0ce…:339)
    at new Tracker.Computation (tracker.js?hash=9f8a0ce…:229)
    at Object.Tracker.autorun (tracker.js?hash=9f8a0ce…:604)
meteor.js?hash=e3f53db…:930 Exception from Tracker recompute function:
meteor.js?hash=e3f53db…:930 TypeError: Cannot read property 'splice' of null
    at http://localhost:3000/packages/blaze.js?hash=9d08dd7dde99958947679827a947134cdf2e4f12:2857:37
    at Object.Tracker.nonreactive (http://localhost:3000/packages/tracker.js?hash=9f8a0cec09c662aad5a5e224447b2d4e88d011ef:631:12)
    at Object.addedAt (http://localhost:3000/packages/blaze.js?hash=9d08dd7dde99958947679827a947134cdf2e4f12:2826:17)
    at Object.addedBefore (http://localhost:3000/packages/observe-sequence.js?hash=7657a09b18583bf0e90cc7b86aa029572767a87f:225:17)
    at http://localhost:3000/packages/diff-sequence.js?hash=15014d7b1e11c05111a386992e684ab1d3cc4158:203:42
    at Array.forEach (native)
    at Function._.each._.forEach (http://localhost:3000/packages/underscore.js?hash=cde485f60699ff9aced3305f70189e39c665183c:149:11)
    at DiffSequence.diffQueryOrderedChanges (http://localhost:3000/packages/diff-sequence.js?hash=15014d7b1e11c05111a386992e684ab1d3cc4158:195:5)
    at diffArray (http://localhost:3000/packages/observe-sequence.js?hash=7657a09b18583bf0e90cc7b86aa029572767a87f:209:3)
    at http://localhost:3000/packages/observe-sequence.js?hash=7657a09b18583bf0e90cc7b86aa029572767a87f:142:9
meteor.js?hash=e3f53db…:930 Exception from Tracker recompute function:
meteor.js?hash=e3f53db…:930 TypeError: Cannot read property '0' of null
    at blaze.js?hash=9d08dd7…:2890
    at Object.Tracker.nonreactive (tracker.js?hash=9f8a0ce…:631)
    at Object.changedAt (blaze.js?hash=9d08dd7…:2882)
    at observe-sequence.js?hash=7657a09…:308
    at Function._.each._.forEach (underscore.js?hash=cde485f…:157)
    at diffArray (observe-sequence.js?hash=7657a09…:294)
    at observe-sequence.js?hash=7657a09…:142
    at Object.Tracker.nonreactive (tracker.js?hash=9f8a0ce…:631)
    at observe-sequence.js?hash=7657a09…:116
    at Tracker.Computation._compute (tracker.js?hash=9f8a0ce…:339)
```

**Comment by prosf** [comment]
Hi zenweasel,
Thanks for the feedback.
I went to checkout a second time with a logged in user but couldn't replicate the error.
Can you please try again (with reaction reset) and tell us in more detail the exact steps you took (if it happens again) so we can reproduce it?

**Comment by lcampanis** [comment]
@aaronjudd @zenweasel 
We just pushed a small update to make sure carriers can be updated when changes take place in Shippo (see EDIT in description)
Let us know if you need anything else to be looked at. 

Best now for RC to take over with the UI and add those values in the CartFlow so that international orders can work for all carriers. This in turn will allow for customs labels to be populated for products. 
Hope it all makes sense. Any help or advice you need, just ping us.
Let us know how ti goes - Removing WIP now.

**Comment by zenweasel** [comment]
@lcampanis so if I do
```
reaction init -b reaction-421-shippo reaction-shippo
cd reaction-shippo
reaction
```

Then login as my admin user
then set the shippo API
then go to checkout I get

```
meteor.js?hash=e3f53db…:930 Exception from Tracker recompute function:
meteor.js?hash=e3f53db…:930 TypeError: Cannot set property 'value' of null
    at ElementAttributesUpdater.update (blaze.js?hash=9d08dd7…:1425)
    at Blaze.View.updateAttributes (blaze.js?hash=9d08dd7…:1590)
    at blaze.js?hash=9d08dd7…:1934
    at Function.Template._withTemplateInstanceFunc (blaze.js?hash=9d08dd7…:3746)
    at blaze.js?hash=9d08dd7…:1932
    at Object.Blaze._withCurrentView (blaze.js?hash=9d08dd7…:2273)
    at viewAutorun (blaze.js?hash=9d08dd7…:1931)
    at Tracker.Computation._compute (tracker.js?hash=9f8a0ce…:339)
    at Tracker.Computation._recompute (tracker.js?hash=9f8a0ce…:358)
    at Object.Tracker._runFlush (tracker.js?hash=9f8a0ce…:523)
```

I feel like this was a bug (not yours) that was fixed in an upstream branch so maybe something just needs to get updated?

**Comment by lcampanis** [comment]
@zenweasel We'll test this now and see if we can find anything and let you know.
@aaronjudd Just realised the PR description wasn't updated. It is now so just look for **EDIT**
Forgot to mention multiple items have not been implemented. 
Might be best if you guys take over to make sure the code is adjusted, then plug the UI, and then we can focus on multiple items. :)

**Comment by lcampanis** [comment]
@zenweasel We have tested it again a multiple times and cannot replicate this on mac or linux, running the exact same command and taking exactly the same steps. The order goes through successfully. 
Here's the log
```
Reaction successfully installed!

To start your Reaction instance, just run:

 cd reaction-shippo
 reaction

lorenzo@lorenzos-MacBook-Pro:~/Sites/artlimes.com/test$ cd reaction-shippo
lorenzo@lorenzos-MacBook-Pro:~/Sites/artlimes.com/test/reaction-shippo$ reaction

Using settings file at settings/dev.settings.json

Setting up plugin imports...

[[[[[ ~/Sites/artlimes.com/test/reaction-shippo ]]]]]

=> Started proxy.
=> Started MongoDB.
cfs:gridfs: updating npm dependencies -- mongodb, gridfs-stream...
15:43:45.915Z  INFO Reaction: Load default data from /private/data/
15:43:46.654Z  INFO Reaction: JobServer started
15:43:46.658Z  WARN Reaction: Skipped loading settings from reaction.json.
15:43:51.392Z  INFO Reaction: Using meteor --settings to create admin user
15:43:53.566Z  WARN Reaction:

   *********************************

    IMPORTANT! DEFAULT ADMIN INFO

    EMAIL/LOGIN: whatrfrr@localhost

    PASSWORD: 8wZEpPTi

   *********************************


15:43:53.656Z  INFO Reaction: Migrating from version 0 -> 2
15:43:53.657Z  INFO Reaction: Running up() on version 1
15:43:53.813Z  INFO Reaction: Running up() on version 2
15:43:53.876Z  INFO Reaction: Finished migrating.
15:43:53.982Z  WARN Reaction: OpenExchangeRates API not configured. Not adding fetchRates job
15:43:53.986Z  WARN Reaction: OpenExchangeRates API not configured. Not adding flushRates job
15:43:53.987Z  INFO Reaction: ::: Adding notification route permissions to default roles
15:43:54.448Z  INFO Reaction: Reaction initialization finished.
=> Started your app.

=> App running at: http://localhost:3000/
15:44:10.418Z  INFO Reaction: ProductSearch collection (re)built successfully.
15:51:52.587Z  WARN Reaction: Failed to get currency exchange rates.
15:52:17.811Z  WARN Reaction: Failed to get currency exchange rates.
15:53:42.969Z  WARN Reaction: Failed to get currency exchange rates.
15:53:59.381Z  WARN Reaction: Failed to get currency exchange rates.
15:55:26.618Z  INFO Reaction: after cart update, call inventory/addReserve
15:55:26.634Z  INFO Reaction: Moving Inventory items from new to reserved
15:55:26.646Z  INFO Reaction: totalRequiredQty 1
15:55:26.647Z  INFO Reaction: availableInventoryQty 19
15:55:26.647Z  INFO Reaction: updating reservation status 1 of 1/1 items.
15:55:26.661Z  INFO Reaction: finished creating 1 new reserved reservations
15:55:26.828Z  INFO Reaction: cart: add variant SMr4rhDFnYvFMtDTX to cartId KmhP9gqR6mc6Qcfjr
15:56:10.370Z  INFO Reaction: cart/copyCartToOrder KmhP9gqR6mc6Qcfjr
15:56:10.457Z  INFO Reaction: Moving Inventory items from reserved to sold
15:56:10.467Z  INFO Reaction: totalRequiredQty 1
15:56:10.468Z  INFO Reaction: availableInventoryQty 1
15:56:10.468Z  INFO Reaction: updating reservation status 1 of 1/1 items.
15:56:10.474Z  INFO Reaction: finished creating 2 new sold reservations
15:56:10.492Z  INFO Reaction: Created orderId YkRhi9D7DgacGwZ8S
15:56:10.774Z  INFO Reaction: Transitioned cart KmhP9gqR6mc6Qcfjr to order YkRhi9D7DgacGwZ8S
15:56:10.893Z  INFO Reaction: orders/sendNotification status: new
15:56:11.905Z  WARN Reaction: Failed to get currency exchange rates.
15:56:12.860Z  WARN Reaction: Failed to get currency exchange rates.
15:56:12.921Z  INFO Reaction: sending notification to user: bJdP8Fo42ogf75R7G
15:56:12.940Z  WARN Reaction: Error: error occured while sending sms [TypeError: Cannot read property 'apiKey' of undefined]
15:56:12.940Z  INFO Reaction: Sending notification to bJdP8Fo42ogf75R7G
15:56:13.006Z  INFO Reaction: sending notification to admin
15:56:13.026Z  WARN Reaction: Error: error occured while sending sms [TypeError: Cannot read property 'apiKey' of undefined]
15:56:13.026Z  INFO Reaction: Sending notification to bJdP8Fo42ogf75R7G
15:56:13.035Z  WARN Reaction: Reaction.Email.getMailUrl() - no email provider configured
15:56:13.050Z ERROR Reaction: Mail not configured
```

**Comment by lcampanis** [comment]
@zenweasel Also - We tested using the Test token from Shippo
image

**Comment by prosf** [comment]
@zeweasel
I merged the branch with the up-to-date release-0.19.0.
Could you please check if the error persists?
Thx

**Comment by aaronjudd** [comment]
I am seeing the same error that @zenweasel has reported.  The steps to replicate are just that you need to place more than one order in the same session. Console of second order attempt always fails trying to load shipping..


Another UI update that might need to be done -> most shipping companies have a requirement that you don't do rate comparison while displaying a carrier/rate.  Ie, can't have FedEx 2 Day and UPS 2 Day displayed, but rather need to have "2 Day - $xxx" as a combined, generic rate.    @lcampanis is it a requirement that you display the various providers, or is this an update that makes sense for your use case as well.

I'll dig into this now and report back, as I'd like to merge this in the 0.19 working branch, so that we can start updating with the necessary UI changes for shipping. This should be merged before we work on the updates related to #1698 and #1404.

**Comment by lcampanis** [comment]
@aaronjudd We're looking into the error now based on your feedback.

Re carrier display - _is it a requirement that you display the various providers, or is this an update that makes sense for your use case as well._ 
No requirement and it makes sense since these can be consolidated into one using the `days` key here https://goshippo.com/docs/reference#rates - Would it be possible for RC to accommodate this change along with the other UI fixes as we need to wrap up the rest?

Understood about merging before the UIX tickets are done - Hopefully we'll have the cart error fixed today and report back. 

Shippo Tracking Updates are coming your way as well (from start - up until delivery) - currently in internal review. We'll be doing Returns in a couple of weeks, once payments are done.

**Comment by lcampanis** [comment]
Hi @aaronjudd 
We spent the entire day just on that error. We can only reproduce the problem `very` rarely, and with no definite way of triggering it. 

It's definitely a problem with how templates are re-rendered in the cart when an Address is Changed/Added/Removed, etc. Or in other words the reactiveness of the Cart. It's a blaze issue after all.

We plugged Shippo in the existing cartFlow, which gets new rates every time a change is done in the cart, just like the other shippingMethods do.
`shipmentQuotes` is rendering those quotes in release 0.19 between 5-10 times (that number of times is not consistent). This in itself seems to be a problem. You would want to render those options only once, but since the cart is reactive and gets updated with new quotes, the render triggers, multiple times and "seems" to falls into an infinite loop.

Here's an idea of what `might` be going on:
1. Any change in the cart will trigger a render
2. Step2 edit/add/remove address `OR` Step3 change carrier option - will trigger a cart re-render (many many times) 
2. In any of those cases, we have to get those quotes and update the cart when we get them. This will again re-render the entire cart, as well as the quotes, causing this infinite loop.

If you could look at how the cart renders shipping quotes it will make more sense. We think that the structure behind the reactiveness of the cart needs to be reviewed by RC asap.

**Comment by aaronjudd** [comment]
@lcampanis ok, I'm going to merge this into the release branch, and I'm going to start working on it from there (as we're trying to also make update to the UI approach in that branch now).

**Pull request merged**